### PR TITLE
issue #28

### PR DESCRIPTION
**Pull request by niruc100**
add logout button to admin.html

Edit: After a few hours I was able to solve both issues. It is now a SPA with Login/Logout integrated. 
Hope this is what you thought it should be.

### CODE REVIEW

**Comment by n2o** [comment]
Thanks, I will review this in a moment.

Could you please use 2 Spaces in HTML? Thanks :-)

**Comment by n2o** [comment]
image
The lower list should not be accessible when not logged in

**Comment by n2o** [comment]
Please add an `alert-info` above the login form to show users that they first need to authenticate:

```html
<div class="alert alert-info" role="alert">
  To access the admin panel, you must first login.
</div>
```

**Comment by n2o** [comment]
With you configuration, it is not possible to logout:

http://localhost:8080/logout

image

File: src/main/java/de/hhu/propra/link/security/SecurityConfiguration.java [hunk 1]
```diff
@@ -11,8 +11,11 @@ protected void configure(HttpSecurity http) throws Exception {
         http.authorizeRequests()
                 .antMatchers("/*/delete").hasRole("ADMIN")
                 .antMatchers("/admin").hasRole("ADMIN");
-        http.formLogin().permitAll();
-        http.logout().permitAll();
+        http.formLogin()
+                .loginPage("/login")
+                .permitAll();
+        http.logout()
```
Comment (n2o) [inline src/main/java/de/hhu/propra/link/security/SecurityConfiguration.java:17]: Logout is not working

File: src/main/resources/templates/main.html [hunk 2]
```diff
@@ -0,0 +1,112 @@
+<!DOCTYPE html>
```
Comment (n2o) [inline src/main/resources/templates/main.html]: Rename this file to index.html

File: src/main/resources/templates/main.html [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+<html lang="en" xmlns:th="http://www.thymeleaf.org" xmlns:sec="http://www.w3.org/1999/xhtml">
+<head>
+    <meta charset="UTF-8">
+    <title>Shorty</title>
+    <link href="../static/node_modules/bootstrap/dist/css/bootstrap.min.css" rel="stylesheet"
+          th:href="@{node_modules/bootstrap/dist/css/bootstrap.min.css}">
+</head>
+<body>
+<div class="container pt-3">
+    <h1 class="text-center">
```
Comment (n2o) [inline src/main/resources/templates/main.html]: Please use 2 spaces to indent HTML Code. I will add an issue to this to add an editorconfig.

**Review by n2o — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 4]
```diff
@@ -33,4 +33,4 @@ out/
 .vscode/
 
 db/data/
-production.env
\ No newline at end of file
+skeleton.env
```
Comment (n2o) [inline .gitignore]: This was intended. skeleton.env is a sample configuration, which should reside in the project. The `production.env`, on the other hand, should never be tracked by Git. So please revert these changes.

Comment (n2o) [reply .gitignore]: And re-add the skeleton.env file please

File: src/test/java/de/hhu/propra/link/controllers/LinkControllerTest.java [hunk 5]
```diff
@@ -82,11 +65,18 @@ void testNewLinkAuthorizedWithoutCsrf() throws Exception {
...
                         .param("url", "http://www.abc.de")
                         .with(csrf())
         ).andExpect(status().is3xxRedirection())
-                .andExpect(redirectedUrl("/admin"));
+                .andExpect(redirectedUrl("/"));
+    }
+
+    @Test
+    void testlogin() throws Exception {
+        mvc.perform(get("/login"))
```
Comment (n2o) [inline src/test/java/de/hhu/propra/link/controllers/LinkControllerTest.java:78]: Please add a test for `logout`

**Review by n2o — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 4 above)
Comment (niruc100) [reply .gitignore]: I'm sorry i guess this happend when i used rename function

**Comment by niruc100** [comment]
Hey,

i tried to fix what I was able to. I can only reproduce your Logout Failure when I perform a get Request with my browser, then the LinkController wants to process my /logout as an abbreviation. If i use the Logoutbutton everything works fine. BTW I am using the following dockercontainer to run shorty:

version: "3.7"
services:
  redis:
    image: redis:alpine
    ports:
      - 6379:6379
    command: redis-server --appendonly yes
    volumes:
      - ./db/data:/data

**Comment by iTitus** [comment]
You can use the blacklist to disable routes from becoming abbreviations

File: build.gradle [hunk 6]
```diff
@@ -28,6 +28,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-security'
+    compile 'org.thymeleaf.extras:thymeleaf-extras-springsecurity5'
```
Comment (n2o) [inline build.gradle]: `compile` is deprecated, please use `implementation` instead (just replace this word)

**Review by n2o — CHANGES_REQUESTED** [review]

File: src/main/resources/templates/index.html [hunk 7]
```diff
@@ -15,11 +15,101 @@ <h1 class="text-center">
...
+        <label for="username">Username</label>
+        <input type="text" id="username" name="username" class="form-control"/>
+      </div>
+      <div class="form-group">
+        <label for="password">Password</label>
+        <input type="password" class="form-control" id="password" name="password"/>
+      </div>
+      <input type="hidden"
+             name="${_csrf.parameterName}"
+             value="${_csrf.token}"/>
```
Comment (iTitus) [inline src/main/resources/templates/index.html]: Thymeleaf should add these csrf fields on its own, provided you use some types of th: tags (like th:action) in the form tag.

**Review by n2o — APPROVED** [review]

**Comment by n2o** [comment]
Thanks!
